Replace the use of StringBuilder with StringBuffer

komugi1211s opened the pull request:
Replaces `StringBuilder` with `StringBuffer` across codebases, which seems to be present in a codebase for quite a while and had similar functionality as `StringBuilder`, but with improved performance as measured by the benchmark done by @SlugFiller.

It's mostly a full-on search-and-replace, but there are few modification to make it compile as well.

<!--
Please target the `master` branch in priority.
PRs can target `3.x` if the same change was done in `master`, or is not relevant there.

Relevant fixes are cherry-picked for stable branches as needed by maintainers.
You can mention in the description if the change is compatible with `3.x`.
-->

lawnjelly commented on core/string/string_buffer.h:
```diff
@@ -35,7 +35,7 @@
 template <int SHORT_BUFFER_SIZE = 64>
 class StringBuffer {
-	char32_t short_buffer[SHORT_BUFFER_SIZE];
+	char32_t short_buffer[SHORT_BUFFER_SIZE] = {};
```
I don't know the `StringBuffer` class, but I would double check this is required.

This is kind of controversial - it might lead to needless extra processing if used in a bottleneck, if not optimized out. Profiling / benchmarking may help, maybe there's no cost in practice but it would be nice to show this first. :man_shrugging: 

Some people often do this kind of thing thinking it is somehow "safer", but this is an illusion. Providing you aren't reading from uninitialized data there's no problem. Personally if it's a small structure and not a bottleneck I'd be more likely to zero. If a bottleneck, no. If a large buffer no (because zeroing commits the pages).

AThousandShips commented on core/string/string_buffer.h:
```diff
@@ -147,6 +148,15 @@ String StringBuffer<SHORT_BUFFER_SIZE>::as_string() {
+	if (buffer.is_empty()) {
+		return String(short_buffer, string_length);
+	} else {
+		return buffer;
```
This needs to be `buffer.substr(0, string_length + 1)` as it might be larger than the string length

komugi1211s replied:
thank you!

komugi1211s commented:
I am sincerely sorry for late response, I've been busy lately with my work.
I've been slowly collecting benchmark and investigating why my previous benchmark came out that way, and I think I have a somewhat enough information now.

I feel like I should've asked someone to supervise my approach before doing this :sweat_smile: I tried my best to make accurate measurement, and see if I can improve performance here and there.

<hr>

The reason why `String::operator+=` was faster than `StringBuffer<>` all around was because the right hand value I used for concatenation was already `String`, not `const char*`. `StringBuffer<>(String)` internally uses while loop to determine the length of string before concatenating, as opposed to `String::operator+=(String)` which just uses `String::length` under the hood.

To confirm my suspicion, I benchmarked both  `StringBuffer<>` and `String+=` using `const char*` and `String`.
string of each size (16, 64, 128, 256, 512) are being concatenated into one `StringBuffer<>` or `String` 1000 times, then repeated that process 1000 times (1000 repeat, 1000 workload).

this was the result before the change:
![perf_before](https://github.com/godotengine/godot/assets/70929713/4d1efbe5-6326-4db6-8e34-3b8a7ee7784d)

I updated `StringBuffer<>(String)` with similar approach to `String+=(String)`, and measured again.

this was the result after change in `StringBuffer<>(String)`:
![perf_after](https://github.com/godotengine/godot/assets/70929713/16a1b2d8-3317-4539-abac-f2f36ee17ba7)

<hr>

next, the performance difference of `StringBuffer<>(const char*)` between debug and release was suspicious. I wasn't sure why it's not gaining as much speed as `String::operator+=(const char*)`, so I looked into it and found out that these lines are the culprit:
https://github.com/godotengine/godot/blob/5c2295ff538312884115c2b7a3aec1e301b8b954/core/string/string_buffer.h#L100-L101

because this constantly dereferences pointer to see if it reached the end of string (instead of relying on the result of `strlen` prior to this section), this for loop wasn't being optimized at all.

This is the compiler output of said line for release build:
![image_before_vectorization](https://github.com/godotengine/godot/assets/70929713/726f03bc-859f-4e54-8980-581abbf14ffb)

replacing line above with simple for-loop using the result of `strlen` improved performance drastically.

this was the result after change in `StringBuffer<>(String)`, and  `StringBuffer<>(const char*)`:
![perf_vector](https://github.com/godotengine/godot/assets/70929713/828e299f-1e14-4f65-85b7-d018d4eaf342)

This was the compiler output after the for-loop change, which is doing what I wanted it to do:
![image_aftor_vectorization](https://github.com/godotengine/godot/assets/70929713/0ab13b0f-85a7-45e2-8e99-17a1460254ff)

komugi1211s committed:
Replaces the use of StringBuilder with StringBuffer.

- Replaces the use of StringBuilder with StringBuffer.
- updates StringBuffer to be more performant on certain cases.

lawnjelly commented:
Great work! :+1:
Just to let you know this is not forgotten, just waiting as I think @reduz may be on holiday, so we can check he is okay with this.

reduz commented:
I am sorry but I just don't see how this can have better performance than StringBuilder, specially if some optimizations are made to StringBuilder that are pending (moving the internal vectors to a LocalVector with a reserve usage). The approach in StringBuilder pretty much does not allocate memory until the end and it does not do unnecesary memory copying.

SlugFiller commented:
@reduz You can try benchmarking. The current benchmarks show a massive difference. Also, even if there was a different yet more efficient approach, it could simply replace `StringBuffer` as well. There would still be no reason to have two different implementations of the same thing.

I could explain the theory of why` StringBuilder` is so significantly slower, as I have many times, but such a theoretical explanation is irrelevant in the face of the practical benchmark data.

lawnjelly commented:
> I could explain the theory of why StringBuilder is so significantly slower, as I have many times, but such a theoretical explanation is irrelevant in the face of the practical benchmark data.

We do try to keep a good explanation of the reasoning for PRs available _from the PR_, simply so it is easier for reviewers to come in and read all the relevant information rather than just effectively a "trust me, this is better".

Afaik the original discussion took place on rocket chat, I'd forgotten the details myself as it was a while ago:
https://chat.godotengine.org/channel/devel?msg=27YwiqFc8ndSnAnes

komugi1211s commented:
@reduz sorry for taking so long to respond! I wished to post a response sooner, but I was busy with my work.

The StringBuffer performs better compared to StringBuilder because the concatenation is performed in a simple loop, allowing for the compiler to optimize. this comes with the cost of possibly allocating excess memory in some situations. I admit that I did not measure the amount of memory allocated for each case, only the speed in a controlled environment. I would like to measure the memory when I have time.

Given that StringBuilder is used to generate the compiled shader code from the visual shader editor, I hope that this provides a better experience to people who like the feature, although I do think that the gain is insignificant aside from this particular case.

If possible, I would love to get some advice on where to go with this PR. I think from the performance perspective this is better suited for general use, but I am not yet familiar with the codebase so I may be missing something important.

AThousandShips commented:
> The `StringBuffer` performs better compared to `StringBuilder` because the concatenation is performed in a simple loop, allowing for the compiler to optimize

Optimize how and what?

komugi1211s commented:
@AThousandShips After re-reading my previous comment I realized that I did not provide any substance on my claim, I apologize. I will try to explain as much as I can

I was thinking about the `append` method that uses for loop to append `const char*`, which can now be replaced with SIMD instruction to process multiple bytes at once, resulting in a better performance. I am not sure if the same level of optimization is being applied to (or can be applied to) StringBuilder, but looking at the benchmark taken in a release build it is unlikely. unfortunately this seems to be only applicable for `const char *` path.

when appending `String` on the other hand, although the concatenation itself uses identical method (simply using `memcpy`), StringBuilder has a slight overhead from copying a string object because it needs to store the copy of it in order to delegate the actual concatenation to the end, which adds up when gathering a lot of small string (which seems to be the original use case of StringBuilder). StringBuffer does not have the same problem.

I hope this answers the question, and let me know if I missed anything.

YYF233333 commented:
Come across this accidentally, it seems this PR has a simillar goal with #97777.

Firstly I have a question:

> I was thinking about the append method that uses for loop to append const char*, which can now be replaced with SIMD instruction to process multiple bytes at once, resulting in a better performance.

Do you mean this?
https://github.com/godotengine/godot/blob/9e6098432aac35bae42c9089a29ba2a80320d823/core/string/string_buffer.h#L100C1-L102C3

I think `StringBuilder` do exactly the same thing.
https://github.com/godotengine/godot/blob/9e6098432aac35bae42c9089a29ba2a80320d823/core/string/string_builder.cpp#L84C1-L86C5

And not relevant but I think both of them can utilize `memcpy` for better performance since they know the length of the c string.

-----
Back to `StringBuffer`, I think it is ill-designed, it is basicly just a `String` with on-stack cache, better renamed to `InlinableString` or simillar. This is a pretty common optimization for `String`, take [this rust ver.](https://crates.io/crates/inlinable_string) as an example. 

`StringBuilder` is a fundementally different thing. What it wants to solve is multiple resize during string concat. E.g. you have three string "aaa", "bbb", "ccc", if you use `+=` it resize twice, but if using `StringBuilder` you only resize once.

The problem of `StringBuilder` is that it is to heavyweight, containing three `Vector`, slow to construct. The solution is to reuse them as much as possible (static/thread_local). 

So in summary, I think `StringBuffer` is another type of string, it is not a *Builder* so we cannot use it to replace `StringBuilder`. The overhead of `StringBuilder` should be optimized, but not go back to `operator +=`.

The inline cache itself is a good thing and maybe we can add that to `String`? Should be a big perf gain in small string concat.

SlugFiller commented:
It is annoying for me to keep having to explain it, but I'll do it one more time. `StringBuilder` isn't "different". It is _wrong_. There is no scenario in which it makes sense to use the algorithm it uses. Now I will explain why.

`StringBuilder` attempts to take a bunch of strings, pre-allocate the buffer into which they are inserted, and then copy them into said buffer. In theory, this seems to make sense. It avoids any need to reallocate the buffer as you go.

Where it goes wrong is in the "take a bunch" bit. It doesn't accept an iterator that returns the strings on the fly. Rather, the strings are added one by one, before being iterated at the generation step. And the way it "adds" a string is by inserting it into a `Vector` (actually, 2 different ones for each add, but that just increases the already existing issue).

The issue is that appending an element to a `Vector` is, in and of itself, just as expensive as enlarging a buffer to be able to copy an extra string into it. This means that just at the "take a bunch" step, before the accumulation even begins, `StringBuilder` has already used the same amount of resources as `StringBuffer` uses in its entire operation.

As a result, every single instruction of the accumulation step is overhead. By default, `StringBuilder` will always, and in every possible scenario, simply be a slower version of `StringBuffer`. There is nothing it can do, faster or otherwise, that `StringBuffer` can't.

I honestly, swear to Carmack, thought this is obvious. But if it wasn't, then here, a full explanation as to why `StringBuilder` uses a suboptimal algorithm. If it's still unclear, I'm open to any clarifying questions.

The only scenario `StringBuilder`'s algorithm makes sense, is if you somehow already have an array or iterator of strings. An example where it might happen in practice, is if you have a pre-made array of strings, and you're constantly replace one in the middle and rebuild the entire concatenation. However, `StringBuilder` doesn't have a method which accepts an array or vector, nor a method for replacing one of the strings in the middle. In other words, there is no way to use it, as it is currently built, in a way that justifies the algorithm it uses.

By contrast, `StringBuffer` is significantly faster, because it only adds each string once, and doesn't have a final accumulation step. In other words, it has 0 overhead, and is the fastest possible algorithm for appending strings one at a time. And the benchmarks confirm this.

P.S. Making the `StringBuilder` vectors `static threadlocal` could help, reducing allocations during the "take a bunch" step. It would come with a cost: You'd only be able to use one `StringBuilder` at a time (in any given thread), making the process bug-prone. For instance, this would not work:
```cpp
StringBuilder str1;
StringBuilder str2;
str1 += 'prefix_';
str2 += 'different_';
if (strs.size() > 0) {
  str1 += strs[0];
  str2 += strs[0];
}
for (int i = 1; i < strs.size(); i++) {
  str1 += '_';
  str2 += '_';
  str1 += strs[i];
  str2 += strs[i];
}
str1 += '_suffix';
str2 += '_and_unique';
String res1 = str1;
String res2 = str2; // If static threadlocal buffers are used, this would be the same as res1, even though it's not supposed to be
```
But this exact same solution could be applied to `StringBuffer` - using a `static threadlocal` buffer to hold the final string, then copying a substring (with a single allocation) as a replacement "accumulation" step. And it would carry the same issue: As soon as you create a new `StringBuffer`, any existing one becomes invalid, as it overwrites the same buffer. At this point, you might as well have a string-building singleton.

YYF233333 commented:
Firstly thanks for your detailed explanation @SlugFiller , I still have some point to discuss.

> Where it goes wrong is in the "take a bunch" bit. It doesn't accept an iterator that returns the strings on the fly. Rather, the strings are added one by one, before being iterated at the generation step. And the way it "adds" a string is by inserting it into a Vector (actually, 2 different ones for each add, but that just increases the already existing issue).

The problem is that iterator doesn't exist in the usecase, typically the string you want to concat are local variables/string literals that are distributed on stack. You need some bookkeeping, a builder, to gather them up. Bookkeeping do has overhead, so we use it only when simple `+=` becomes bottleneck.

> The issue is that appending an element to a Vector is, in and of itself, just as expensive as enlarging a buffer to be able to copy an extra string into it. This means that just at the "take a bunch" step, before the accumulation even begins, StringBuilder has already used the same amount of resources as StringBuffer uses in its entire operation.

Sorry I forget we are still using `Vector` in master. `Vector::push_back` resize to `size()+1` is frustrating, but luckily we have `LocalVector` which does normal x2 resize.  I optimize this in my PR.

> By contrast, StringBuffer is significantly faster, because it only adds each string once, and doesn't have a final accumulation step. In other words, it has 0 overhead, and is the fastest possible algorithm for appending strings one at a time. And the benchmarks confirm this.

Problem is, `+=` does exactly the same thing, and the benchmark proves this, the performance of `StringBuffer` and `+=` are at same level. Then why use `StringBuffer` instead of direct concat?

The only usecase of `StringBuffer` in `VariantParser::get_token` shows that. What it actually want is a stack array to store small string. If the string is too long fallback to normal `+=`. This is good and we can use this stratergy in more place, but it should be named as `InlinableString`, not `StringBuilder`/`StringBuffer`.

So if the profile shows that `StringBuilder` is slow than direct concat, feel free to fallback to `+=`, but that should be case by case,  not find-and-replace in a batch.

YYF233333 commented:
> For compiling 300 GLES3 shaders in one frame master spends about 200 ms in StringBuilder, #97777 spends about 90 ms (almost all of that time is in `as_string()`), and this PR spends about 28 ms (all of that time is in append).

That's expected and means probably the bookkeeping overhead is larger than resize overhead and we should use `+=` instead. Theoretically in this case I think time costed low to high:
- `+=` with a stack buffer == `StringBuffer`
- `+=` only
- `static StringBuilfer`
- `StringBuilfer` only

> Do you mind testing the same scene in https://github.com/godotengine/godot/pull/97777 with the code from this PR to compare performance?

I will if I have time.

P.S. This PR reminds me that I'm trapped by the old implementation. We should directly use a `String` as buffer since we need to return one, whatever else buffer we use is adding another `memcpy`. Will correct this ASAP.

Again, I'm not against this class. I'm against its name and replacing all `StringBuilder`.

SlugFiller commented:
> You need some bookkeeping, a builder, to gather them up. Bookkeeping do has overhead, so we use it only when simple `+=` becomes bottleneck.

And my point was that the bookkeeping is _equally expensive_ as `+=`. Except the bookkeeping requires you to do a collection step later, and `+=` does not. Therefore, it is better to do `+=` _instead_ of the bookkeeping. 

> Sorry I forget we are still using `Vector` in master. `Vector::push_back` resize to `size()+1` is frustrating, but luckily we have `LocalVector` which does normal x2 resize. I optimize this in my PR.

That, in and of itself, is not the issue. It merely contributes to the issue. The issue is that appending to the `Vector` or `LocalVector` is as expensive as expanding a buffer.

> Problem is, `+=` does exactly the same thing, and the benchmark proves this, the performance of `StringBuffer` and `+=` are at same level. Then why use `StringBuffer` instead of direct concat?

You're suggesting replacing `StringBuffer` with just `String`. This is... fair. It largely depends on how `String` is implemented. I believe, however, that `String` has additional copy-on-write logic, since it's supposed to be reusable and immutable, whereas `StringBuffer` does not. But don't take my word for it, I didn't dive into the `String` source code yet.

> So if the profile shows that `StringBuilder` is slow than direct concat, feel free to fallback to `+=`, but that should be case by case, not find-and-replace in a batch.

It's not just the profile, although the profile is pretty conclusive. Logically, `StringBuilder` will _always_ be slower than `+=`, in every single possible case. This is an inevitable result of its design. This is why a search-and-replace is appropriate.

It's like replacing bubble sort with quicksort. There's no conceivable condition under which the prior would be preferable, other than maybe code simplicity.  And `StringBuilder`'s algorithm is anything but simple.

SlugFiller commented:
> I am struggling to understand the appeal of keeping `StringBuilder`. I understand that it does reduce the amount of reallocation during the gathering phase

...but adds an equal or larger amount of reallocations during the append phase, thereby removing any semblance of having an advantage or purpose.

> but is there a guarantee that the `const char *` kept in the `StringBuilder` is alive at the time when `StringBuilder::as_string` is called?

I think there isn't. But in order for your example code to actually create garbage, you need to thrash the stack in between the `append` and the call to `to_string`. One way to do it is to call another function, which uses several stack variables. That would corrupt the result.

If this was Rust or Swift, an escaping or linked lifetime guarantee could be required from the `const char *` to prevent it. But, generally, the `const char *` version is intended purely for constant strings which are located in the program code, and therefore are guaranteed to last the entire lifetime of the process.

Ivorforce commented:
I ran some of my own benchmarks against `StringBuilder`, `StringBuffer` and simply `String +=`, taking into account what I've learned from working with the classes over the past few days.

Before starting, I patched `StringBuffer` to its (i believe) optimal state, because it currently does some unnecessary allocations, and we cannot draw any final conclusions with half unfinished implementations. The optimizations are mainly 
- Build string directly in buffer, instead of building elsewhere and copying over
- `LocalVector` instead of Vector
- One vector instead of 3

(I tested all 3 optimizations iteratively top to bottom, and each contribute to better scores).

<details>
<summary>string_builder.h</summary>

```c++
/**************************************************************************/
/*  string_builder.h                                                      */
/**************************************************************************/
/*                         This file is part of:                          */
/*                             GODOT ENGINE                               */
/*                        https://godotengine.org                         */
/**************************************************************************/
/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
/*                                                                        */
/* Permission is hereby granted, free of charge, to any person obtaining  */
/* a copy of this software and associated documentation files (the        */
/* "Software"), to deal in the Software without restriction, including    */
/* without limitation the rights to use, copy, modify, merge, publish,    */
/* distribute, sublicense, and/or sell copies of the Software, and to     */
/* permit persons to whom the Software is furnished to do so, subject to  */
/* the following conditions:                                              */
/*                                                                        */
/* The above copyright notice and this permission notice shall be         */
/* included in all copies or substantial portions of the Software.        */
/*                                                                        */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
/**************************************************************************/

#ifndef STRING_BUILDER_H
#define STRING_BUILDER_H

#include "core/string/ustring.h"
#include "core/templates/vector.h"

#include <core/templates/local_vector.h>

class StringBuilder {
	struct Item {
		String s;
		const char *ptr;
		size_t ptr_len;
	};

	uint32_t string_length = 0;

	LocalVector<Item> strings;
public:
	StringBuilder &append(const String &p_string);
	StringBuilder &append(const char *p_cstring);

	_FORCE_INLINE_ StringBuilder &operator+(const String &p_string) {
		return append(p_string);
	}

	_FORCE_INLINE_ StringBuilder &operator+(const char *p_cstring) {
		return append(p_cstring);
	}

	_FORCE_INLINE_ void operator+=(const String &p_string) {
		append(p_string);
	}

	_FORCE_INLINE_ void operator+=(const char *p_cstring) {
		append(p_cstring);
	}

	_FORCE_INLINE_ int num_strings_appended() const {
		return strings.size();
	}

	_FORCE_INLINE_ uint32_t get_string_length() const {
		return string_length;
	}

	String as_string() const;

	_FORCE_INLINE_ operator String() const {
		return as_string();
	}
};

#endif // STRING_BUILDER_H

```
</details>

<details>
<summary>string_builder.cpp</summary>

```c++
/**************************************************************************/
/*  string_builder.cpp                                                    */
/**************************************************************************/
/*                         This file is part of:                          */
/*                             GODOT ENGINE                               */
/*                        https://godotengine.org                         */
/**************************************************************************/
/* Copyright (c) 2014-present Godot Engine contributors (see AUTHORS.md). */
/* Copyright (c) 2007-2014 Juan Linietsky, Ariel Manzur.                  */
/*                                                                        */
/* Permission is hereby granted, free of charge, to any person obtaining  */
/* a copy of this software and associated documentation files (the        */
/* "Software"), to deal in the Software without restriction, including    */
/* without limitation the rights to use, copy, modify, merge, publish,    */
/* distribute, sublicense, and/or sell copies of the Software, and to     */
/* permit persons to whom the Software is furnished to do so, subject to  */
/* the following conditions:                                              */
/*                                                                        */
/* The above copyright notice and this permission notice shall be         */
/* included in all copies or substantial portions of the Software.        */
/*                                                                        */
/* THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,        */
/* EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF     */
/* MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. */
/* IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY   */
/* CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT,   */
/* TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE      */
/* SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.                 */
/**************************************************************************/

#include "string_builder.h"

#include <string.h>

StringBuilder &StringBuilder::append(const String &p_string) {
	if (p_string.is_empty()) {
		return *this;
	}

	strings.push_back({p_string, nullptr, 0});
	string_length += p_string.length();

	return *this;
}

StringBuilder &StringBuilder::append(const char *p_cstring) {
	size_t len = strlen(p_cstring);
	if (len == 0) {
		return *this;
	}

	strings.push_back({String(), p_cstring, len});
	string_length += len;

	return *this;
}

String StringBuilder::as_string() const {
	if (string_length == 0) {
		return "";
	}

	String string;
	string.resize(string_length);
	char32_t *buffer_ptr = string.ptrw();

	for (uint32_t i = 0; i < strings.size(); i++) {
		const auto& item = strings[i];

		if (!item.ptr) {
			// Godot string
			const String &s = item.s;

			memcpy(buffer_ptr, s.ptr(), s.length() * sizeof(char32_t));
			buffer_ptr += s.length();
		} else {
			const char *s = item.ptr;

			for (int32_t j = 0; j < item.ptr_len; ++j, ++buffer_ptr, ++s) {
				*buffer_ptr = *s;
			}
		}
	}

	return string;
}
```
</details>

Here's the code I ran:

<details>
<summary>test_main.cpp</summary>

(templating should optimize away `StrBuilder`, just adding it for convenience).

```c++
struct StrBuilder {
	String x;

	void append(const String& s) {
		x += s;
	}

	String as_string() {
		return x;
	}
};

template <typename Builder>
void test(const String &s, int runs, int count) {
	for (int r = 0; r < runs; ++r) {
		Builder b;
		for (int i = 0; i < count; ++i) {
			b.append(s);
		}
		int l = b.as_string().length();
	}
}

int test_main(int argc, char *argv[]) {
	String s = String("A").repeat(1);
	int runs = 100;
	int count = 200;

	test<StringBuilder>(s, runs, count);
	auto t0 = std::chrono::high_resolution_clock::now();
	test<StringBuilder>(s, runs, count);
	auto t1 = std::chrono::high_resolution_clock::now();
	std::cout << std::chrono::duration_cast<std::chrono::microseconds>(t1 - t0).count() << "us\n";

	test<StringBuffer<>>(s, runs, count);
	t0 = std::chrono::high_resolution_clock::now();
	test<StringBuffer<>>(s, runs, count);
	t1 = std::chrono::high_resolution_clock::now();
	std::cout << std::chrono::duration_cast<std::chrono::microseconds>(t1 - t0).count() << "us\n";

	test<StrBuilder>(s, runs, count);
	t0 = std::chrono::high_resolution_clock::now();
	test<StrBuilder>(s, runs, count);
	t1 = std::chrono::high_resolution_clock::now();
	std::cout << std::chrono::duration_cast<std::chrono::microseconds>(t1 - t0).count() << "us\n";
}
```
</details>

I ran the test with 4 configurations, differing in `String` length and `append` count. The reason is that I suspected `StringBuffer` to perform the best with many small appends, `StringBuilder` to perform best with few big appends, and `+=` performing about on-par with `StringBuffer` for big strings but worse for small strings.

Indeed, this is exactly what we see:
```
(Each StringBuilder, StringBuffer, String +=)

10000 length, 5 runs, 3 appends
10us
64us
17us

20 length, 100 runs, 20 appends
88us
51us
68us

40 length, 200 runs, 20 appends
168us
173us
148us

1 length, 100 runs, 200 appends
1253us
284us
472us
```

I hope it's clear what the reason for this is: 
- `StringBuilder` performs badly on very short strings because the book-keeping overhead overtakes the benefit of storing the strings.
- `StringBuffer` performs badly on very long strings because it allocates more than is needed, and re-allocates several times. 
- `String +=` is never considerably better than the others. 
- The handover happens at about `40`-length strings.
- Both perform badly, up to a factor of 6x, when used for strings that are too short or too long.

I think this is enough data for me to make a judgement call.

I propose the following: 
- Rename `StringBuffer` to `ShortStringBuilder` and `StringBuilder` to `LongStringBuilder` (or something along those lines).
- `StringBuilder` should be addressed ASAP for its inefficiencies.
- Repetitive short-string appends in `StringBuilder` (such as in [shader_gles3.cpp](https://github.com/godotengine/godot/blob/893bbdfde8ad1f94fb4e6db246ff7075569396ea/drivers/gles3/shader_gles3.cpp#L215-L220) should be avoided. Instead, a longer string should be appended at once.
- For strings for which the components are known at compile-time, [concatenate_strings](https://github.com/godotengine/godot/pull/99929) should be used.
- For multiple appends at once, a function `extend` should be added to both `StringBuilder` and `StringBuffer`, to reduce buffer growth overhead. It should be statically evaluated, modeled after [concatenate_strings](https://github.com/godotengine/godot/pull/99929).
- `char *` arguments should be replaced with `StrRange` arguments, to avoid unnecessary `strlen` calls, as proposed in https://github.com/godotengine/godot/pull/99806.

SlugFiller commented:
> * `StringBuffer` performs badly on very long strings

"Performs badly" is a misphrasing here. The difference is in microseconds. It's on the level of measurement error, and is roughly 20 times less than the difference in the many appends version.

To be complete, you should probably test many appends long length, or at least increase the run count until the run time is measured in seconds, not in microseconds.

Or you can flip the stop condition, and measure number of runs that can be completed before a second elapses. This should somewhat reduce measurement error, although overhead from repeatedly getting the current time might take over in the case of a very short run.

At any rate, even if we take these measurements to be perfectly accurate, with no errors introduced from resolution limitation of the system clock, being able to earn 50us for the risk of losing 1000us is not a proper tradeoff. It does not justify the use of `StringBuilder`, even in its optimized form.

Ivorforce commented:
> "Performs badly" is a misphrasing here. The difference is in microseconds. It's on the level of measurement error, and is roughly 20 times less than the difference in the many appends version.

Hrm, you are right, unfortunately this does cast a shadow on my conclusions. 

But i agree with clay that if we went there, to try to totally replace StringBuilder, StringBuffer should be renamed to StringBuilder, instead of renaming references.

Also, StringBuffer using a short string cache is not suitable for building strings, since the data has to be copied over later anyhow. This should be addressed before attempting to propose a switch.

Ivorforce commented:
Urgh, I just now realized that `String` already grows in power-of-2 steps. This is likely the reason it's doing decently well for concatenation in the first place.

With that in the picture as well, I wonder why `String +=` is even slower than `StringBuffer` in the first place, because all `StringBuffer` does is some additional overhead for resizing by power-of-2 (logic that is repeated in the string resize...).

clayjohn commented:
> Urgh, I just now realized that `String.cow_data` already grows in power-of-2 steps. This is likely the reason it's doing decently well for concatenation in the first place.
> 
> With that in the picture as well, I wonder why `String +=` is even slower than `StringBuffer` in the first place, because all `StringBuffer` does is some additional overhead for resizing by power-of-2 (logic that is repeated in the string resize...).

String has a lot more bookeeping steps, look at the number of instructions in order to get to the Po2 resize logic. StringBuffer does the resize up front because it knows its going to need to resize a lot. Basically you skip all the string and CoW logic
